example/calculate_pi/calculate_pi.cpp

```diff
@@ -37,11 +37,57 @@ namespace calculate_pi {
 
 void SetInOrOut(MeshBlockData<Real> *rc) {
   auto pmb = rc->GetBlockPointer();
+
+  ParArrayND<Real> v;
+  const auto &radius = pmb->packages.Get("calculate_pi")->Param<Real>("radius");
+
+  // If we're using sparse variables, check if we need to allocate the sparse id before
+  bool const use_sparse = pmb->packages.Get("calculate_pi")->Param<bool>("use_sparse");
+  if (use_sparse) {
+    int const ndim = pmb->pmy_mesh->ndim;
+    auto &bs = pmb->block_size;
+    // check if block falls on radius.
+    Real coords[4][2] = {{bs.x1min, bs.x2min},
+                         {bs.x1min, bs.x2max},
+                         {bs.x1max, bs.x2min},
+                         {bs.x1max, bs.x2max}};
+
+    bool inside = false;
+    bool outside = false;
+
+    for (auto i = 0; i < 4; i++) {
+      Real const rsq = coords[i][0] * coords[i][0] + coords[i][1] * coords[i][1];
+      if (rsq < radius * radius) {
+        inside = true;
+      } else {
+        outside = true;
+      }
+    }
+
+    // If part of the block falls inside and other parts fall outside, then we need to
+    // allocate the sparse id before computing on it. If it's wholly inside or outside,
+    // then there's no need for expansion.
+
+    // TODO(agaspar): Change this to only allocate sparse ids that fall on the radius.
+    // Currently we're allocating all sparse ids that fall outside the circle at all. if
+    // (!inside || !outside) {
+    if (inside && !outside) {
+      // std::cout << "Out: " << pmb->gid << std::endl;
+      return;
+    }
+
+    // std::cout << "In: " << pmb->gid << std::endl;
+
+    rc->AllocSparseID("in_or_out", 0);
+    v = rc->Get("in_or_out", 0).data;
+  } else {
+    v = rc->Get("in_or_out").data;
+  }
```

Comment:
Yeah, I wasn't happy with this code either. It wasn't quite clear and it actually gave the wrong answer. I fixed it now and try to make it more understandable. The idea is that we check if all four corners of a block fall completely outside the circle. If so, we don't need to allocate `in_or_out` on that block, because it would be all zeros (and when we sum up areas, if it's not allocated, we simply skip it).

Of course, right now in this PR, we always allocate sparse fields on all blocks, so it won't make a difference in this PR, but once we have true sparse fields implemented, this will be a neat little demonstration.

Reply:
Cool. Yeah this makes sense and should be pretty neat.